tokenizer/src/main/scala/anorm/TokenizedStatement.scala

```diff
@@ -0,0 +1,139 @@
+package anorm
+
+import scala.util.{ Failure, Try, Success }
+
+trait Show {
+  def show: String
+}
+
+private[anorm] sealed trait StatementToken
+private[anorm] case class StringToken(value: String) extends StatementToken
+private[anorm] case object PercentToken extends StatementToken
+
+private[anorm] case class TokenGroup(
+  /** Already prepared tokens, not requiring to rewrite placeholder. */
+  prepared: List[StatementToken],
+
+  /** Optional placeholder (name), after already prepared tokens */
+  placeholder: Option[String])
+
+private[anorm] case class TokenizedStatement(
+  /** Token groups */
+  tokens: List[TokenGroup],
+
+  /** Binding names of parsed placeholders */
+  names: List[String])
+
+private[anorm] object TokenizedStatement {
+  import java.util.StringTokenizer
+  import scala.collection.JavaConverters.enumerationAsScalaIteratorConverter
+  import scala.reflect.macros.Context
+  import scala.language.experimental.macros
+  
+  /** Returns empty tokenized statement. */
+  lazy val empty = TokenizedStatement(Nil, Nil)
+
+  /** String interpolation to tokenize statement. */
+  def stringInterpolation[T](parts: Seq[String], params: Seq[T with Show]): (TokenizedStatement, Map[String, T]) = macro tokenizeImpl[T]
+
+  /** Tokenization macro */
+  def tokenizeImpl[T: c.WeakTypeTag](c: Context)(parts: c.Expr[Seq[String]], params: c.Expr[Seq[T with Show]]): c.Expr[(TokenizedStatement, Map[String, T])] = 
```

Comment:
Allow to tokenize at compile time for Anorm string interpolation (no extra load at runtime).